socs/common/camera.py

```diff
@@ -0,0 +1,392 @@
+            to_load = min(len(self._deque), self.recent)
+            for idx in range(to_load):
+                pos = -(idx + 1)
+                file = self._deque[pos][0]
+                self._deque[pos] = (file, self.reader(file))
+
+    def store(self, data):
+        now = datetime.now(tz=timezone.utc)
+        path = self._media_path(now)
+        self.writer(data, path, **self.writer_opts)
+        self.prune()
+        if self.recent > 0:
+            self._deque.append((path, data))
+        else:
+            self._deque.append((path, None))
+
+        # Clear stale recent entry
+        if len(self._deque) > self.recent and self.recent > 0:
+            pos = -(self.recent + 1)
+            file, buffer = self._deque[pos]
+            del buffer
+            self._deque[pos] = (file, None)
+
+        # Update latest symlink
+        link = os.path.join(self.dir, f"latest.{self.suffix}")
+        if os.path.exists(link):
+            os.remove(link)
+        os.symlink(path, link)
+
+    def fetch_recent(self):
+        result = list()
+        to_fetch = min(len(self._deque), self.recent)
+        for idx in range(to_fetch):
+            pos = -(to_fetch + idx)
+            result.append(self._deque[pos])
+        return result
+
+    def fetch_index(self, index):
+        return self._deque[index]
+
+    def prune(self):
+        while len(self._deque) >= self.max_files:
+            file, buffer = self._deque.popleft()
+            del buffer
+            self._remove(file)
+
+    def _remove(self, path):
+        if not os.path.isfile:
+            raise RuntimeError(f"{path} does not exist, cannot remove")
+        os.remove(path)
+
+    def _media_path(self, dt):
+        dstr = self._dt_to_iso(dt)
+        return os.path.join(self.dir, f"{self.root}_{dstr}.{self.suffix}")
+
+    def _media_time(self, path):
+        file = os.path.basename(path)
+        pat = re.compile(f"{self.root}_(.*)\\.{self.suffix}")
+        mat = pat.match(file)
+        iso = mat.group(1)
+        return self._iso_to_dt(iso)
+
+    def _dt_to_iso(self, dt):
+        return dt.isoformat(sep="T", timespec="seconds")
+
+    def _iso_to_dt(self, iso):
+        return datetime.datetime.fromisoformat(iso)
+
+
+class MotionDetector:
+    """Class to process images in sequence and look for changes.
+
+    This uses the (stateful) helper tools from OpenCV to detect when an
+    image contains changes from previous ones.
+
+    Args:
+        blur (int):  The odd number of pixels for blurring width
+        threshold (int):  The grayscale threshold (0-255) for considering
+            image changes in the blurred images.
+        dilation (int):  The dilation iterations on the thresholded image.
+        min_frac (float):  The minimum fraction of pixels that must change
+            to count as a detection.
+        max_frac (float):  The maximum fraction of pixels that can change
+            to still count as a detection (rather than a processing error).
```

Comment:
Can you add these arguments as optional kwargs to the agent? We may want to change these if the camera FOV, room size, etc cause motion detection to fail.